Dev

RobertRosca opened the pull request:
Pretty simple package, with a silly name: Venv in Parent IPykernel.

Working with venv's and jupyter can be a bit awkward since you have to activate the venv, then run `python3 -m ipykernel install --user --name venvkernel-name`, then in all the notebooks select that kernel, then before you commit the notebook remember to change it back to the default `python3` as not everybody will have the same kernel/be using venvs.

The standard `python3` kernel is just:

```
{
 "argv": [
  "/usr/bin/python3",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```

Which launches ipykernel with the system `python3`. This package overwrites the default kernel with:

```
{
 "argv": [
  "/usr/bin/python3",
  "-m",
  "vip_ipykernel_launcher",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```

Where the `vip_ipykernel_launcher` performs a search from the current directory up to the root directory, globbing for a `.venv/bin/python3` or `venv/bin/python3` along the way.

If it finds a match for a `python3` executable in a `venv` directory, it then passes the subsequent arguments to it (i.e. `-m ipykernel -f {connection_file}`), which in the end replicates the standard kernel launch process.

Then you can install this kernel and launch a notebook in a project with a `.venv` in it, and that notebook will run with the correct python environment.

Anybody up for reviewing this? @tmichela @takluyver @antarcticrainforest

RobertRosca committed:
Add readme

RobertRosca committed:
Add initial repository files

RobertRosca committed:
Copy ipykernel kernelspec

RobertRosca committed:
Add modifications

RobertRosca committed:
Use setup.py instead

Poetry is overkill when we only have one dependecy

RobertRosca committed:
Update gitignore

RobertRosca committed:
Add clarifying comments

RobertRosca committed:
Adjust how mod argument is used

RobertRosca committed:
Update argparse descriptions

RobertRosca committed:
Add vip ipykernel launcher

RobertRosca committed:
Add attribution

RobertRosca committed:
Sort imports

RobertRosca committed:
Add resources logos

RobertRosca committed:
Rename install app

RobertRosca committed:
Add tests

RobertRosca committed:
Add note about logging

RobertRosca committed:
Update dependencies

RobertRosca committed:
Update readme

RobertRosca committed:
Add pytest-cov to test requirements

RobertRosca committed:
Add tests

RobertRosca committed:
Add publish to pypi

RobertRosca committed:
Add codecov

RobertRosca committed:
Expand todo

RobertRosca commented on src/vip_ipykernel/kernelspec.py:
```diff
@@ -0,0 +1,204 @@
+    #  When installing the ViP IPykernel, the first `-m` module call points to
+    #  our `vip_ipykernel_launcher`, and the second module call points to the
+    #  desired ipykernel launcher module
+    arguments = [executable, '-m', 'vip_ipykernel.vip_ipykernel_launcher', '-m', mod, '-f', '{connection_file}']
```
Here's where I added `'-m', 'vip_ipykernel.vip_ipykernel_launcher'` to 'wrap' the ipykernel launcher call.

RobertRosca commented on src/vip_ipykernel/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,51 @@
+ANCHOR = Path(Path.cwd().anchor)
+
+
+def venv_search(prefix: Path = Path('.')) -> Path:
+    prefix = prefix.absolute()
+
+    found_venvs = []
+
+    for venv in VENV_NAMES:
+        found_venvs.extend(prefix.glob(f"{venv}/bin/python3"))
+
+    if any(found_venvs):
+        #  If there are multiple venvs just return the first one
+        return found_venvs[0].absolute()
+    elif prefix == ANCHOR:
+        return Path(sys.executable)
+    else:
+        return venv_search(prefix=prefix.parent)
```
Search function is from me.

RobertRosca commented on src/vip_ipykernel/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,51 @@
+
+    Instance(logging.Logger)
+
+    args[0] = str(venv_search())
```
`args[0]` is set to the path to the venv python3 executable

RobertRosca commented on src/vip_ipykernel/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,51 @@
+
+    #  TODO: I want to use the jupyter logger to print this off but can't figure
+    #  out how to do that, @takluyver can you help with this?
+    print(f"Starting venv kernel with args: {args}")
```
Kinda crappy, I'd like to log this with the Jupyter Notebook/Lab logger (as per the comment above) but don't know how to do that :disappointed:

RobertRosca committed:
Sort imports

RobertRosca committed:
Fix pytest-cov version

RobertRosca committed:
Add resources dir to package data

RobertRosca committed:
Bump version

RobertRosca committed:
Remove old logging attempt

RobertRosca committed:
Resolve path to system python3

RobertRosca committed:
Run test with shutil.which instaed of sys.executable

RobertRosca committed:
Set version to alpha, use get version in setup.py

RobertRosca committed:
Expand readme

cydanil commented on README.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+This just says "Run use `python3` to run `ipykernel_launcher` with an argument
````
```suggestion
This just says "Run using `python3` to run `ipykernel_launcher` with an argument
```

cydanil commented on src/vip_ipykernel/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,46 @@
+        found_venvs.extend(prefix.glob(f"{venv}/bin/python3"))
+
+    if any(found_venvs):
+        #  If there are multiple venvs just return the first one
```
Perhaps this ought to be highlighted in the readme

RobertRosca replied:
Yes good point. In those kind of situations it'd be nice to give some feedback to users, but there's no (as far as I can tell) simple mechanism to raise kernel initialisation warnings that built into Jupyter Notebook/Lab.

RobertRosca replied:
Added a note about this to the readme

cydanil reviewed (commented):
sick!

RobertRosca committed:
Fix typo

Co-authored-by: Cyril Danilevski <cydanil@gmail.com>

RobertRosca committed:
Clarify venv priority order

RobertRosca committed:
Clarify tests with comments

RobertRosca committed:
Add notebook tests

RobertRosca committed:
Depend on ipykernel, patch it

RobertRosca committed:
Adjust tests for new patching

RobertRosca committed:
Move launcher entry point

RobertRosca committed:
Use the launcher module correctly

RobertRosca committed:
Move venv search out of launcher

RobertRosca committed:
Rename tests

RobertRosca committed:
Update gitignore

RobertRosca committed:
Adjust formatting

RobertRosca committed:
Expand python test versions

RobertRosca committed:
Minor fixes and cleanup

RobertRosca committed:
Update badge formatting

ebadkamil commented:
yes indeed @cydanil @RobertRosca

RobertRosca committed:
Space out VENV_NAMES

RobertRosca committed:
Update tests

- Replace python 3.x with 3.9
- Add caching
- Stop installing pytest/flake8 manually

RobertRosca committed:
Bump version

takluyver commented on src/vip_ipykernel/__init__.py:
```diff
@@ -0,0 +1,28 @@
+    found_venvs = []
+
+    for venv in VENV_NAMES:
+        found_venvs.extend(prefix.glob(f"{venv}/bin/python3"))
```
It doesn't look like there's any actual glob pattern involved? Might be clearer to just check if the exact path exists:

```python
path = prefix / venv / 'bin' / 'python3'
if path.is_file():
    return path.absolute()
```

RobertRosca replied:
Ah yes good point, originally I had a reason to use glob which escapes me now so I'll change it, thanks

takluyver commented on src/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,26 @@
+import subprocess
+import sys
+
+from vip_ipykernel import venv_search
+
+if __name__ == '__main__':
+    # Remove the CWD from sys.path while we load stuff.
+    # This is added back by InteractiveShellApp.init_path()
+    if sys.path[0] == '':
+        del sys.path[0]
```
N.B. `ipykernel_launcher` imports ipykernel after this sys.path check - that's the point of having a separate `ipykernel_launcher` module. It avoids ipykernel failing to even launch because someone has made `random.py` or something in the CWD.

RobertRosca replied:
Aha, yes I didn't fully understand why this was there to be honest. Given that this, in the end, just calls `ipykernel_launcher`, I take it that it would be fine to remove the sys.path check?

takluyver replied:
It's less of an issue than with ipykernel, because you import so much less stuff, but anything you do import (like pathlib) could be shadowed by files in the cwd. So the `_launcher` workaround  removes that from `sys.path` before any imports, and then the ipykernel machinery adds it back later (once most of its own imports are done).

RobertRosca replied:
Ooooh I get it now, huh yeah that's an interesting issue, I'll leave it as it is for now then, thank you for the explanation

takluyver replied:
If you're leaving this module around, you probably want to move the `from vip_ipykernel` import after this line. Otherwise this code isn't really doing anything.

takluyver commented on src/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,26 @@
+    if sys.path[0] == '':
+        del sys.path[0]
+
+    args = sys.argv
```
```suggestion
    args = sys.argv.copy()
```

Doesn't really matter here, but it's a good idea to avoid modifying globals if possible.

takluyver commented on src/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,26 @@
+
+    #  TODO: I want to use the jupyter logger to print this off but can't figure
+    #  out how to do that, @takluyver can you help with this?
+    print(f"Starting venv kernel with args: {args}")
```
I don't think anything has set up logging - you haven't even imported any Jupyter/IPython modules at this point, as far as I can see. You could call e.g. `logging.basicConfig()` yourself, but I don't think there's any major advantage over plain `print()`.

RobertRosca replied:
Basically I just wanted, for no real reason, to print this output to the Jupyter Notebook/Lab logging. Currently it looks like:

```
[I 13:54:14.648 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 13:54:14.709 NotebookApp]

    To access the notebook, open this file in a browser:
        file:///home/roscar/.local/share/jupyter/runtime/nbserver-3635158-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/?token=5e4191cd8032e03cf7618509400265a9e6d67e56b98f1eb3
     or http://127.0.0.1:8888/?token=5e4191cd8032e03cf7618509400265a9e6d67e56b98f1eb3
[I 13:54:56.902 NotebookApp] Kernel started: 1fc13264-3b04-4015-85b9-6452750217ae, name: python3
Starting venv kernel with args: ['/usr/bin/python3.8', '-m', 'ipykernel_launcher', '-f', '/home/roscar/.local/share/jupyter/runtime/kernel-1fc13264-3b04-4015-85b9-6452750217ae.json']
```

And I'd like:

```
[I 13:54:14.648 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 13:54:14.709 NotebookApp]

    To access the notebook, open this file in a browser:
        file:///home/roscar/.local/share/jupyter/runtime/nbserver-3635158-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/?token=5e4191cd8032e03cf7618509400265a9e6d67e56b98f1eb3
     or http://127.0.0.1:8888/?token=5e4191cd8032e03cf7618509400265a9e6d67e56b98f1eb3
[I 13:54:56.902 NotebookApp] Kernel started: 1fc13264-3b04-4015-85b9-6452750217ae, name: python3
[I 13:54:SS.sss NotebookApp]  Starting venv kernel with args: ['/usr/bin/python3.8', '-m', 'ipykernel_launcher', '-f', '/home/roscar/.local/share/jupyter/runtime/kernel-1fc13264-3b04-4015-85b9-6452750217ae.json']
```

I've tried using `logging.basicConfig()` to get the current logger, but that doesn't pick up the one the server is using, also tried figuring out how to access the logger instance traitlet (https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelbase.py#L68) but I've never really used or looked into traitlets so they're still pretty mysterious to me

tbh you're right since `print` achieves the same thing, all that's missing is the timestamp so if this is too much effort then that's fine since it really doesn't matter much, I was just curious if there's an easy way to use the logger directly

takluyver replied:
Your code is running in a child process of the notebook server, so there's no way to access the server's logging machinery directly, and AFAIK Jupyter doesn't expose any special interfaces for kernels to do logging. If you want the timestamp, you could set up the logging framework in your kernel process, but I probably wouldn't bother if it was me.

RobertRosca replied:
Yeah that's what I thought, print is good enough for this so I'll leave it as it is

takluyver commented on src/vip_ipykernel_launcher.py:
```diff
@@ -0,0 +1,26 @@
+    #  out how to do that, @takluyver can you help with this?
+    print(f"Starting venv kernel with args: {args}")
+
+    subprocess.run(args)
```
Potentially bad advice: for something like this I might use [os.execv](https://docs.python.org/3/library/os.html#os.execv). It's a somewhat scary low-level interface, but it goes into the new program directly under the current pid, rather than creating a new child process and then waiting for it to finish.

I think we fixed things in Jupyter so it works correctly anyway (e.g. signals go to the process group rather than the individual process), so this *shouldn't* matter. It's more of a neat trick.

RobertRosca replied:
Aha, didn't even know that was a thing, the current subprocess run approach works fine but it does make sense to start it under the current pid since otherwise you'll always have two 'kernels' running, with the vip-ipykernel one just... doing nothing after it starts up the actual kernel. I'll make the change now, thanks :smile:

RobertRosca replied:
Hmm I'm mildly confused by the purpose of this, and the docs don't really explain it:

> The “v” variants are good when the number of parameters is variable, with the arguments being passed in a list or tuple as the args parameter. In either case, the **arguments to the child process should start with the name of the command being run**, but this is not enforced.

The first element of the arguments passed to the process are not actually passed to it, they seem to be discarded (e.g. `os.execv("/usr/bin/echo", ["1", "2"])` will echo `2` but not `1`). So I just pass the full args list instead of `args[1:]`, since element 0 is the path to the python executable, and it'll be ignored anyway.

So `os.execv(args[0], args)` works fine but I'm wondering if you know/could explain why the first argument is dropped and why it should "start with the name of the command being run"?

takluyver replied:
I think it's just one of those mismatches between the high-level interfaces we're used to and lower-level interfaces. It's a strong convention that argv[0] is the script or executable running in that process, and higher level APIs (like bash or the subprocess module) assume that to avoid duplication. But at the low level, it's just a convention, and you can break it if you want to.

RobertRosca replied:
Aha so it's just a convention thing, I was wondering if the argument went anywhere or did... something

takluyver commented on tests/test_nb.py:
```diff
@@ -0,0 +1,103 @@
+        assert str(tmp_path) in jupyter_client.kernelspec.find_kernel_specs()['python3']
+
+        #  The kernel should be running with the same python that is running the tests
+        assert sys.executable in spec["argv"][0]
```
```suggestion
        assert spec["argv"][0] == sys.executable
```

Maybe?

RobertRosca replied:
Ah yes I think this is in a different comment somewhere, here it uses `in` because depending on the system and how the paths are resolved `sys.executable` could be `python3` but `spec["argv"][0]` may be `python3.8`. For example:

```ipython
Python 3.8.6 (default, Sep 30 2020, 04:00:38) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sys

In [2]: from pathlib import Path

In [3]: sys.executable
Out[3]: '/usr/bin/python3'

In [4]: Path(sys.executable).resolve()
Out[4]: PosixPath('/usr/bin/python3.8')
```

So it's possible for sys.executable to be `/usr/bin/python3` but for the (resolved) path in the spec to be `/usr/bin/python3.8`

takluyver commented on tests/test_nb.py:
```diff
@@ -0,0 +1,103 @@
+    from vip_ipykernel.kernelspec import install as install_vip
+    import vip_ipykernel.kernelspec
+
+    os.chdir(tmp_path)
```
Use the [monkeypatch fixture](https://docs.pytest.org/en/latest/monkeypatch.html) to avoid chaning the cwd for later tests?

RobertRosca replied:
Aha, clever! Also didn't know that was a thing, thanks :D

takluyver commented on tests/test_nb.py:
```diff
@@ -0,0 +1,103 @@
+        #  Should be using the vip_ipykernel module
+        assert "vip_ipykernel" in spec["argv"][2]
+
+        #  Test the ViP kernel falling back to base python
+        pytest.main([
+            "--verbose",
+            "--nbval",
+            "notebook-std.ipynb",
+        ])
```
Does `pytest.main()` ever call `sys.exit()`? If so, it might just stop after the first one of these tests that gets run.

RobertRosca replied:
Oops, screwed this up but for a different reason. `pytest.main` doesn't raise an exception if it fails it just returns the exit code of the tests, so this should actually do `assert pytest.main([...]) == 0`, I'll fix this and the other call lower down, thanks!

takluyver commented on tests/test_nb.py:
```diff
@@ -0,0 +1,103 @@
+
+    shutil.copy(NB_FILE_VIP, ".")
+
+    #  Create venv in the temporary directory
+    subprocess.run([
+        "python3",
+        "-m",
+        "venv",
+        ".venv"
+    ])
```
There's a [venv.create()](https://docs.python.org/3/library/venv.html#venv.create) function - might give better error reporting if things go wrong.

RobertRosca replied:
Aha, another thing I didn't know existed :joy: awesome I'll use this instead, thanks! Who'da thunk this silly kernel thing would end up being so educational :smile:

RobertRosca committed:
Standardise python module calls

RobertRosca committed:
No need for glob, use `is_file` to check for python3

RobertRosca committed:
Copy argv before modifying it

RobertRosca committed:
Replace `subprocess.run` call with `os.execv`

RobertRosca committed:
Use `monkeypatch.chdir` instead of `os.chdir`

RobertRosca committed:
Fix nested calls to pytest

RobertRosca committed:
Use call to venv module instead of subprocess

RobertRosca committed:
Sort imports

RobertRosca committed:
Add .vscode to gitignore

RobertRosca commented:
Think I covered all of your comments @takluyver, marked most of the comments as resolved but left a few unresolved for now, if you have anything to add to the remaining ones then feel free to do so when you have time or to mark them as resolved. For like the 8th time thank you for reviewing this, it was really useful!

RobertRosca committed:
Standardise python calls for builds

takluyver commented:
Just gave one last comment; otherwise I think it's fine, but I don't have a button to mark them as resolved. :-)

RobertRosca committed:
Move import to after sys.path manipulation

RobertRosca committed:
Update readme

RobertRosca committed:
Bump version to 1.0.0

RobertRosca committed:
Use minimum versions for install requires

RobertRosca committed:
Change lifecycle to active

RobertRosca committed:
Add dependabot integration and constraints
